command/agentproxyshared/cache/lease_cache.go

```diff
@@ -267,15 +336,25 @@ func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 		inflight.remaining.Inc()
 		defer inflight.remaining.Dec()
 
-		c.inflightCache.Set(id, inflight, gocache.NoExpiration)
-		idLock.Unlock()
+		c.inflightCache.Set(staticSecretCacheId, inflight, gocache.NoExpiration)
+		idLockStaticSecret.Unlock()
```

Comment:
Why don't we need to hold idLock any longer?  Doesn't cacheStaticSecret, called below, rely on a lock being held to allow safe concurrent updates?

Reply:
Admittedly, I mostly copied this logic from the dynamic secret flow, but my understanding is:
- We check if the request is in the in flight cache
- If it is, we wait until it's completed
- If it's not, we hold a lock while we add it

So, we don't need the lock as the other clients are in this flow:
```
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-inflight.ch:
```

and we only need to lock the inflight cache, essentially

Reply:
Ok, ty, I hadn't understood that behaviour.  I still think your introduction of cacheStaticSecret changes things in a way that it's unsafe to release the lock.  Suppose there's one goroutine that takes the else branch and creates a newInflightRequest.  Now two other goroutines come in, and block waiting for inflight.ch to be closed.  First goroutine finishes and closes the channel.  Now both new goroutines are allowed to proceed, and they can contend when trying to append to the token slice.

Reply:
Hmm, I thought about this some more and I'm not sure that I see the problem. You're only allowed to proceed if there's not already an inflight request with the same `staticSecretCacheId`. If those two differ, then there's no issues with them updating the list of tokens, as it'll be different lists of tokens.

Maybe I'm misunderstanding or missing a code path, though.

Reply:
The issue with updating the list of tokens is that if two routines are doing it at the same time, they can't safely update the list of tokens without some sort of lock or threadsafe structure.  I believe I gave a scenario above where that's possible: 3 goroutines for the same request path, first one creates the channel, upon channel close both other goroutines are unblocked and allowed to proceed down the code path that modifies the token slice.

Reply:
Chatted with Nick over zoom and he explained the code path to me. Added locking to writes, as a result, and resolving this thread!